Treat UnevaluatedExpr specially in CodePrinter.doprint

bjodah opened the pull request:
#### References to other Issues or PRs
https://github.com/sympy/sympy/issues/21991


#### Brief description of what is fixed or changed
When printing e.g. C or Fortran code for the an expression such as `abs(exp(x + (y + z)))`, (where `UnevaluatedExpr` is used to emit parenthesis around (y+z) for cases where the non-associativity floating point addition is an issue), the printer would emit `re(...)` nodes which aren't supported by the printer.

This PR replaces `UnevaluatedExpr` by instances which passes `is_real` queries through.

#### Other comments


#### Release Notes
<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

bjodah committed:
Add kwarg UnevaluatedClass to create_expand_pow_optimization

oscarbenjamin commented on sympy/codegen/rewriting.py:
```diff
@@ -294,6 +295,11 @@ def create_expand_pow_optimization(limit, *, base_req=lambda b: b.is_symbol):
+    UnevaluatedClass : subclass of UnevaluatedExpr
+         By default UnevaluatedExpr is used to prevent automatic simplification
+         of the expanded powers. For certain purposes, you may want to use a
+         subclass thereof instead (which e.g. propagates .is_real by overloading
```
This seems like an unnecessarily general solution to a very specific problem that should perhaps be solved in a different way.

Why is it necessary to control the assumptions like this?

oscarbenjamin replied:
Maybe the same effect could be better achieved by a specific rewrite.

bjodah replied:
What prompted me to do this was to avoid `re(...)` being emitted by the printer. This was the smallest patch I could come up with, but of course it would be nice with something less intrusive.

A rewrite pass which removes `re(...)` if it's argument can be shown to be real is one way, the rewriting pass then needs to manually "jump" into the arg of UnevaluatedExpr instanceses, does it not?

oscarbenjamin replied:
Maybe or what causes the `re` in the first place? Maybe the printers should be fixed or something.

bjodah replied:
The printer cannot determine that the argument is real, perhaps best illustrated with this example:
```python
>>> r, s, t = symbols('r s t', real=True)
>>> ccode(abs(exp(r+s+t)))
'exp(r + s + t)'
>>> ccode(abs(exp(r+UnevaluatedExpr(s+t))))
'// Not supported in C:\n// re\nexp(r + re(s + t))'
```

oscarbenjamin replied:
So maybe the printer should look inside an UnevaluatedExpr

bjodah replied:
Hmmm, yes, that's probably the better fix. Will see what I can come up with.

bjodah committed:
Merge branch 'master' into ccode-unevaluated-expr

bjodah committed:
Handle UnevaluatedExpr in CodePrinter.doprint

bjodah committed:
Adapt pre-processing in doprint to lambdify

bjodah committed:
Allow None to be passed to doprint (used by CSE routine)

bjodah commented:
OK, so I added a pre-processing step in the CodePrinter instead, what do you think of this approach?

oscarbenjamin commented:
I don't know this part of the codebase so well. This seems better then the previous version though.

I keep thinking there must be a better way than needing to invoke a new class but I can't think what it is.

bjodah committed:
Avoid introducing a class to handle UnevaluatedExpr in CodePrinter

oscarbenjamin commented on sympy/printing/codeprinter.py:
```diff
@@ -67,6 +69,15 @@ def __init__(self, settings=None):
+            return expr.replace(re, lambda arg: arg if isinstance(
+                arg, UnevaluatedExpr) and arg.args[0].is_real else re(arg))
+        else:
+            return type(expr)(self._handle_UnevaluatedExpr(item) for item in expr)
```
Why does this need to handle non-Basic?

bjodah replied:
I just pushed a commit which moves this non-Basic handling into lambdify's internal class.

bjodah commented:
Tests for lambdify passes those through lambdarepr printer (which is
questionable in my opinion, but that's where we are at).

On Sat, Sep 11, 2021, 18:16 Oscar Benjamin ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In sympy/printing/codeprinter.py
> <https://github.com/sympy/sympy/pull/21993#discussion_r706631231>:
>
> > @@ -67,6 +69,15 @@ def __init__(self, settings=None):
>          if not hasattr(self, 'reserved_words'):
>              self.reserved_words = set()
>
> +    def _handle_UnevaluatedExpr(self, expr):
> +        if expr is None:
> +            return expr
> +        elif isinstance(expr, Basic):
> +            return expr.replace(re, lambda arg: arg if isinstance(
> +                arg, UnevaluatedExpr) and arg.args[0].is_real else re(arg))
> +        else:
> +            return type(expr)(self._handle_UnevaluatedExpr(item) for item in expr)
>
> Why does this need to handle non-Basic?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/pull/21993#pullrequestreview-751929913>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADWUMAUQKS2IXIEHBIEJWDUBN6GLANCNFSM5DFEO3UQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

bjodah committed:
Handle nesting in _EvaluatorPrinter instead of CodePrinter

bjodah committed:
handle non-SymPy types in lambdastr too

bjodah committed:
Selectively convert list -> List in CodePrinter.doprint

bjodah committed:
Move import of codegen.pynodes.List

oscarbenjamin commented:
Is the List change related to the other changes?

Also are you finished with this or are you still working on it?

bjodah committed:
Merge branch 'master' into ccode-unevaluated-expr

bjodah commented:
@oscarbenjamin, this is good to go from my point of view (barring my latest merge of master causes tests to fail).

Yes, the List change is related: in order for e.g. `_handle_UnevaluatedExpr` to be able to call `.replace`, we cannot have e.g. python-list instances at that point. But for example the tests of the OctavePrinter (in test_octave.py) explicitly exercises printing of python lists. Therefore, I introduced `List` and also added what I hope are some sensible tests of such a class. I noted that `sympify` by default changes python tuple into Tuple instances, but leaves list instances be. By having a "code-printer specific List-class" (`sympy.codegen.pynodes.List`), we can continue to support this behavior without touching non-code-printing related code.

So the design at this point is quite influenced by the constraints set by our existing test suite. Perhaps there's a more elegant solution, but right now, this is the best I've got...

oscarbenjamin commented:
Okay, great. I've set to auto-merge. Thanks
